File: bloock-bridge/go/internal/test/acceptance_test.go
package test

import (
	"math"
	"math/rand"
	"os"
	"strconv"
	"testing"

	bloock "github.com/bloock/go-bridge/client"
	"github.com/stretchr/testify/assert"
)

func GetSdk() bloock.Client {
	apiKey := os.Getenv("API_KEY")
	apiHost := os.Getenv("API_HOST")
	client := bloock.NewClient(apiKey)
	client.SetApiHost(apiHost)
	return client
}

func TestAcceptance(t *testing.T) {
	sdk := GetSdk()

	t.Run("Basic test E2E", func(t *testing.T) {
		record, err := sdk.NewRecordFromString(randHex(64))
		assert.NoError(t, err)
		records := []*bloock.Record{record}

		receipt, err := sdk.SendRecords(records)
		assert.NoError(t, err)
		assert.Greater(t, len(receipt), 0)
		assert.NotEqual(t, bloock.RecordReceipt{}, receipt[0])

		anchor, err := sdk.WaitAnchor(receipt[0].Anchor)
		assert.NoError(t, err)
		assert.Equal(t, receipt[0].Anchor, anchor.Id)

		proof, err := sdk.GetProof(records)
		assert.NoError(t, err)

		root, err := sdk.VerifyProof(proof)
		assert.NoError(t, err)

		timestamp, err := sdk.ValidateRoot(root, bloock.ListOfNetworks().BloockChain)
		assert.NoError(t, err)
		assert.Greater(t, timestamp, uint64(0))

		timestamp, err = sdk.VerifyRecords(records)
		assert.NoError(t, err)
		assert.Greater(t, timestamp, uint64(0))
	})

	t.Run("Test send records invalid record input wrong char", func(t *testing.T) {
		record, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aG")
		assert.NoError(t, err)

		records := []*bloock.Record{record}
		_, err = sdk.SendRecords(records)

		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test send records invalid record input missing chars", func(t *testing.T) {
		record1, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa")
		assert.NoError(t, err)
		record2, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994")
		assert.NoError(t, err)

		records := []*bloock.Record{record1, record2}
		_, err = sdk.SendRecords(records)

		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test send records invalid record input wrong start", func(t *testing.T) {
		record1, err := sdk.NewRecordFromHash("0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa")
		assert.NoError(t, err)
		record2, err := sdk.NewRecordFromHash("0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994bb")
		assert.NoError(t, err)

		records := []*bloock.Record{record1, record2}

		_, err = sdk.SendRecords(records)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test send records empty record input", func(t *testing.T) {
		res, err := sdk.SendRecords([]*bloock.Record{})
		assert.NoError(t, err)
		assert.Nil(t, res)
	})

	t.Run("Test get anchor non existing anchor", func(t *testing.T) {
		_, err := sdk.GetAnchor(666666666666666666)
		assert.Error(t, err)
		assert.Equal(t, bloock.AnchorNotFoundError, err.Error())
	})

	t.Run("Test wait anchor non existing anchor", func(t *testing.T) {
		_, err := sdk.WaitAnchor(666666666666666666, 3000)
		assert.Error(t, err)
		assert.Equal(t, bloock.WaitAnchorTimeoutError, err.Error())
	})

	t.Run("Test get proof invalid record input wrong char", func(t *testing.T) {
		record, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aG")
		assert.NoError(t, err)
		records := []*bloock.Record{record}

		_, err = sdk.GetProof(records)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test get proof invalid record input missing chars", func(t *testing.T) {
		record1, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa")
		assert.NoError(t, err)
		record2, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994")
		assert.NoError(t, err)

		records := []*bloock.Record{record1, record2}

		_, err = sdk.GetProof(records)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test get proof invalid record input wrong start", func(t *testing.T) {
		record1, err := sdk.NewRecordFromHash("0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa")
		record2, err := sdk.NewRecordFromHash("0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994bb")

		records := []*bloock.Record{record1, record2}

		_, err = sdk.GetProof(records)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test get proof non-existant leaf", func(t *testing.T) {
		record, err := sdk.NewRecordFromHash("0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdee")
		assert.NoError(t, err)

		records := []*bloock.Record{record}

		_, err = sdk.GetProof(records)
		assert.Error(t, err)
		assert.Equal(t, bloock.RecordNotFoundError, err.Error())
	})

	t.Run("Test verify records invalid record input wrong char", func(t *testing.T) {
		record, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aG")
		assert.NoError(t, err)
		records := []*bloock.Record{record}

		_, err = sdk.VerifyRecords(records, bloock.ListOfNetworks().BloockChain)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test verify records invalid record input missing chars", func(t *testing.T) {
		record1, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa")
		assert.NoError(t, err)
		record2, err := sdk.NewRecordFromHash("e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994")
		assert.NoError(t, err)

		records := []*bloock.Record{record1, record2}

		_, err = sdk.VerifyRecords(records, bloock.ListOfNetworks().BloockChain)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test verify records invalid record input wrong start", func(t *testing.T) {
		record1, err := sdk.NewRecordFromHash("0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa")
		record2, err := sdk.NewRecordFromHash("0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994bb")

		records := []*bloock.Record{record1, record2}

		_, err = sdk.VerifyRecords(records, bloock.ListOfNetworks().BloockChain)
		assert.Error(t, err)
		assert.Equal(t, bloock.InvalidRecordError, err.Error())
	})

	t.Run("Test verify records non-existant leaf", func(t *testing.T) {
		record, err := sdk.NewRecordFromHash("0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdee")

		records := []*bloock.Record{record}

		_, err = sdk.VerifyRecords(records, bloock.ListOfNetworks().BloockChain)
		assert.Error(t, err)
		assert.Equal(t, bloock.RecordNotFoundError, err.Error())
	})
}

func randHex(length int) string {
	maxlength := 8
	min := math.Pow(16, math.Min(float64(length), float64(maxlength))-1)
	max := math.Pow(16, math.Min(float64(length), float64(maxlength))) - 1
	n := int((rand.Float64() * (max - min + 1)) + min)
	r := strconv.Itoa(n)
	for len(r) < length {
		r += randHex(length - maxlength)
	}
	return r
}

File: languages/js/test/acceptance.test.ts
import { BloockClient, Network, Record } from "../dist/index"
import { InvalidNumberOfRecords, InvalidRecordError, RecordNotFoundError, WaitAnchorTimeoutError } from "../src/errors"

function getSdk(): BloockClient {
    const apiKey = process.env['API_KEY'] || ''
    const apiHost = process.env['API_HOST'] || ''
    let client = new BloockClient(apiKey, apiHost)
    client.setApiHost(apiHost)
    return client
}

function randHex(len: number): string {
    const maxlen = 8
    const min = Math.pow(16, Math.min(len, maxlen) - 1)
    const max = Math.pow(16, Math.min(len, maxlen)) - 1
    const n = Math.floor(Math.random() * (max - min + 1)) + min
    let r = n.toString(16)
    while (r.length < len) {
        r = r + randHex(len - maxlen)
    }
    return r
}

describe('Acceptance Tests', () => {
    test('test_basic_e2e', async () => {
        jest.setTimeout(120000)

        const sdk = getSdk()

        const records = [await Record.FromString(randHex(64))]

        const sendReceipt = await sdk.sendRecords(records)
        if (!sendReceipt) {
            expect(false)
            return
        }

        await sdk.waitAnchor(sendReceipt[0].anchor)

        // Retrieving record proof
        const proof = await sdk.getProof(records)
        const root = await sdk.verifyProof(proof)
        const timestamp = await sdk.validateRoot(root, Network.BLOOCK_CHAIN)
        const isValid = await sdk.verifyRecords(records)

        expect(timestamp).toBeGreaterThan(0)
        expect(isValid).toBeGreaterThan(0)
    })

    test('test_send_records_invalid_record_input_wrong_char', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aG')
        ]

        await expect(sdk.sendRecords(records)).rejects.toEqual(InvalidRecordError)
    })

    test('test_send_records_invalid_record_input_missing_chars', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa'),
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994')
        ]

        await expect(sdk.sendRecords(records)).rejects.toEqual(InvalidRecordError)
    })

    test('test_send_records_invalid_record_input_wrong_start', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa'),
            await Record.FromHash('0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994bb')
        ]

        await expect(sdk.sendRecords(records)).rejects.toEqual(InvalidRecordError)
    })

    test('test_send_records_empty_record_input', async () => {
        const sdk = getSdk()

        const result = await sdk.sendRecords([])

        expect(result).toEqual([])
    })

    test('test_wait_anchor_non_existant_anchor', async () => {
        jest.setTimeout(5000)
        const sdk = getSdk()
        await expect(sdk.waitAnchor(66666666, 3000)).rejects.toEqual(WaitAnchorTimeoutError)
    })

    test('test_get_proof_invalid_record_input_wrong_char', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aG')
        ]

        await expect(sdk.getProof(records)).rejects.toEqual(InvalidRecordError)
    })

    test('test_get_proof_invalid_record_input_missing_chars', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa'),
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994')
        ]

        await expect(sdk.getProof(records)).rejects.toEqual(InvalidRecordError)
    })

    test('test_get_proof_invalid_record_input_wrong_start', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa'),
            await Record.FromHash('0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994bb')
        ]

        await expect(sdk.getProof(records)).rejects.toEqual(InvalidRecordError)
    })

    test('test_get_proof_empty_record_input', async () => {
        const sdk = getSdk()

        await expect(sdk.getProof([])).rejects.toEqual(InvalidNumberOfRecords)
    })

    test('test_get_proof_non_existant_leaf', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef')
        ]

        await expect(sdk.getProof(records)).rejects.toEqual(RecordNotFoundError)
    })

    test('test_verify_records_invalid_record_input_wrong_char', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aG')
        ]

        await expect(sdk.verifyRecords(records, Network.BLOOCK_CHAIN)).rejects.toEqual(InvalidRecordError)
    })

    test('test_verify_records_invalid_record_input_missing_chars', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa'),
            await Record.FromHash('e016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994')
        ]

        await expect(sdk.verifyRecords(records, Network.BLOOCK_CHAIN)).rejects.toEqual(InvalidRecordError)
    })

    test('test_verify_records_invalid_record_input_wrong_start', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994aa'),
            await Record.FromHash('0xe016214a5c4abb88b8b614a916b1a6f075dfcf6fbc16c1e9d6e8ebcec81994bb')
        ]

        await expect(sdk.verifyRecords(records, Network.BLOOCK_CHAIN)).rejects.toEqual(InvalidRecordError)
    })

    test('test_verify_records_empty_record_input', async () => {
        const sdk = getSdk()
        await expect(sdk.verifyRecords([], Network.BLOOCK_CHAIN)).rejects.toEqual(InvalidNumberOfRecords)
    })

    test('test_verify_records_non_existant_leaf', async () => {
        const sdk = getSdk()
        const records = [
            await Record.FromHash('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef')
        ]

        await expect(sdk.verifyRecords(records, Network.BLOOCK_CHAIN)).rejects.toEqual(RecordNotFoundError)
    })
})
